rust/lance-linalg/src/simd/u8.rs

```diff
@@ -0,0 +1,372 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! `u8x8`, 8 of `u8` values
+
+use std::fmt::Formatter;
+
+#[cfg(target_arch = "aarch64")]
+use std::arch::aarch64::*;
+#[cfg(target_arch = "loongarch64")]
+use std::arch::loongarch64::*;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
+#[cfg(target_arch = "loongarch64")]
+use std::mem::transmute;
+use std::ops::{Add, AddAssign, Mul, Sub, SubAssign};
+
+use super::{Shuffle, SIMD};
+
+/// 16 of 8-bit `u8` values.
+#[allow(non_camel_case_types)]
+#[cfg(target_arch = "x86_64")]
+#[derive(Clone, Copy)]
+pub struct u8x16(pub __m128i);
+
+/// 16 of 32-bit `f32` values. Use 512-bit SIMD if possible.
+#[allow(non_camel_case_types)]
+#[cfg(target_arch = "aarch64")]
+#[derive(Clone, Copy)]
+pub struct u8x16(pub uint8x16_t);
+
+impl u8x16 {
+    #[inline]
+    pub fn bit_and(self, mask: u8) -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            Self(_mm_and_si128(self.0, _mm_set1_epi8(mask as i8)))
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            Self(vandq_u8(self.0, vdupq_n_u8(mask)))
+        }
+        #[cfg(target_arch = "loongarch64")]
+        unsafe {
+            Self(lasx_xvfand_b(self.0, mask))
+        }
+    }
+
+    #[inline]
+    pub fn right_shift_4(self) -> Self {
+        #[cfg(target_arch = "x86_64")]
+        unsafe {
+            let shifted = _mm_srli_epi16(self.0, 4);
+            let mask = _mm_set1_epi8(0x0F);
+            Self(_mm_and_si128(shifted, mask))
+        }
+        #[cfg(target_arch = "aarch64")]
+        unsafe {
+            Self(vshrq_n_u8::<4>(self.0))
+        }
+        #[cfg(target_arch = "loongarch64")]
+        unsafe {
+            Self(lasx_xvfrsh_b(self.0, 4))
```

Comment:
huh you figured out how to use longarch?

Reply:
lol no, no way to test it, let me remove all loongarch code for u8x16